File: Session6/C++/GenerateParenthesis.cpp
https://leetcode.com/problems/generate-parentheses/

class Solution {        
    
    void parenthesis(vector<string>&res, string& curr, int n, int open, int close)
    {
        if(curr.size()==n*2)
        {
            res.push_back(curr);
            return;
        }
        
        if(open<n)
        {
            curr.push_back('(');
            parenthesis(res, curr, n, open+1, close);
            curr.pop_back();
        }
        
        if(close < open)
        {
            curr.push_back(')');
            parenthesis(res, curr, n, open, close+1);
            curr.pop_back();
        }
        
    }
    
public:
    vector<string> generateParenthesis(int n) {
        
        vector<string> res;
        string curr="";
        
        parenthesis(res, curr, n, 0, 0);
        
        return res;
        
    }
};

File: Session6/Java/GenerateParenthesis.java
https://leetcode.com/problems/generate-parentheses/

class Solution {
    
    void parenthesis(int n, String curr, int open, int close, List<String> res)
    {
        if(curr.length()==2*n)
        {
            res.add(curr);
            return;
        }

        if(open<n)
        {
            parenthesis(n, curr+'(', open+1, close, res);
        }
        if(close<open)
        {
            parenthesis(n, curr+')', open, close+1, res);
        }
        
    }
    
    public List<String> generateParenthesis(int n) {
        
        List<String> res = new ArrayList<String>();
        parenthesis(n, "", 0, 0, res);
        return res;
        
    }
}